### PR TITLE
torch.optim.lbfgs - added box constraint and line search methods(back…

### PR DESCRIPTION
**Pull request by ChangYong-Oh**
`torch.optim.lbfgs `
- added box constraint 
- added line search methods(backtracking, goldstein, weak_wolfe)
- Box constraint is imposed via line search(line search is conducted under constraints)
- Box constraint is specified by the similar way in scipy fmin_l_bfgs_b
- Line search method choice is given by string(originally, it seems to be going to be given by function)
  

`torch.autograd._function.reduce.Prod class`
- if prod is taken over some tensor containing zero, this returns `inf`, because the original implementation is done by `[product of all elements]/[an element]`.  All these cases are taken care of separately. Still, `tensor.prod()` returns `inf` if there is only one element in a tensor.

This should be fixed.

### CODE REVIEW

File: torch/autograd/_functions/reduce.py [hunk 1]
```diff
@@ -46,13 +46,44 @@ def forward(self, input):
...
-            return grad_input.div(input)
+            zero_loc = (input==0).nonzero()
+            if zero_loc.dim() == 0:
+                grad_input = grad_output.new(self.input_size).fill_(self.result)
+                return grad_input.div(input)
+            elif zero_loc.size()[0] > 1:
+                return grad_output.new(self.input_size).fill_(0)
+            else:
+                grad_input = grad_output.new(self.input_size).fill_(0)
+                indexing_tuple = tuple(zero_loc[0].numpy())
```
Comment (fmassa) [inline torch/autograd/_functions/reduce.py]: This won't work in the GPU

Comment (ChangYong-Oh) [inline torch/autograd/_functions/reduce.py]: Yes you are right.

This will be fixed as
indexing_tuple = tuple((zero_loc[0].cpu() if zero_loc.is_cuda else zero_loc[0]).numpy())
after receiving other feedbacks.

Thanks!

**Comment by apaszke** [comment]
I took your prod fixes and refactored them slightly so that they reuse some intermediate results. Can you please remove this change from this PR? I'll try to get someone who knows line search methods to review your LBFGS changes soon. Thanks!

**Comment by hassec** [comment]
Is there anything one can do to push this along?
IMHO LBFGS would greatly benefit from having these line search functions implemented...

**Comment by soumith** [comment]
@hassec we are blocked on finding competent reviewers for this stuff, who know the relevant literature and can verify the code implemented

**Comment by hassec** [comment]
@soumith thanks for the quick feedback.
Mentioning in #6564 that there is simply a review needed might be useful to find someone?

**Comment by fehiepsi** [comment]
@soumith @hassec I implemented strong Wolfe line search in #8824 with enough references. You might want to take a look if you are interested in. :)

**Comment by SsnL** [comment]
I'll review this in the coming days.

**Comment by weiyangfb** [comment]
@SsnL

**Comment by matthieuheitz** [comment]
What is the state of this PR ?
This L-BFGS-B would be a great addition to Pytorch !

**Comment by ianwilliamson** [comment]
Wondering what the status of this pull request is. Would be great to have constraint options.

**Comment by soumith** [comment]
@matthieuheitz @ianwilliamson would either of you be willing to review the PR for correctness against the papers? that's what it's blocked on at the moment.

**Comment by GeoffNN** [comment]
Is anyone looking at this?

**Comment by ezyang** [comment]
No not at the moment. We need someone to review the PR for correctness (math), and it also needs some rebasing.

File: torch/optim/lbfgs.py [hunk 2]
```diff
@@ -239,3 +252,136 @@ def step(self, closure):
...
+        deriv = 0.0
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            deriv += torch.sum(p.grad.data * d[offset:offset + numel].resize_(p.size()))
+            offset += numel
+        assert offset == self._numel()
+        return deriv
+
+    def _max_alpha(self, d):
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: Reference for Cauchy point computation: equation 16.71 in Numerical Optimization 2nd Edition by Nocedal and Wright.

File: torch/optim/lbfgs.py [hunk 3]
```diff
@@ -239,3 +252,136 @@ def step(self, closure):
...
+        offset = 0
+        max_alpha = float('inf')
+        for p, bnd in zip(self._params, self._bounds):
+            numel = p.numel()
+            l_bnd, u_bnd = bnd
+            p_grad = d[offset:offset + numel].resize_(p.size())
+            if l_bnd is not None:
+                from_l_bnd = ((l_bnd-p.data)/p_grad)[p_grad<0]
+                min_l_bnd = torch.min(from_l_bnd) if from_l_bnd.numel() > 0 else max_alpha
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: This line takes the minimal alpha across all the directions. This would imply that if the point is near the boundary, the next step would be constrained to remain very close. I would expect instead to do different step size in each directions, see 16.71 mentioned in previous comment.

File: torch/optim/lbfgs.py [hunk 4]
```diff
@@ -239,3 +252,136 @@ def step(self, closure):
...
+
+    def _backtracking(self, closure, d):
+        # 0 < rho < 0.5 and 0 < w < 1
+        rho = 1e-4
+        w = 0.5
+
+        original_param_data_list = self._copy_param()
+        phi_0 = closure().data[0]
+        phi_0_prime = self._directional_derivative(d)
+        alpha_k = 1.0
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: Unlike the `_goldstein` and `_weak_wolfe`, this is not invoking `_max_alpha`.

File: torch/optim/lbfgs.py [hunk 5]
```diff
@@ -239,3 +252,136 @@ def step(self, closure):
...
+            numel = p.numel()
+            l_bnd, u_bnd = bnd
+            p_grad = d[offset:offset + numel].resize_(p.size())
+            if l_bnd is not None:
+                from_l_bnd = ((l_bnd-p.data)/p_grad)[p_grad<0]
+                min_l_bnd = torch.min(from_l_bnd) if from_l_bnd.numel() > 0 else max_alpha
+            if u_bnd is not None:
+                from_u_bnd = ((u_bnd-p.data)/p_grad)[p_grad>0]
+                min_u_bnd = torch.min(from_u_bnd) if from_u_bnd.numel() > 0 else max_alpha
+            max_alpha = min(max_alpha, min_l_bnd, min_u_bnd)
```
Comment (vincentqb) [inline torch/optim/lbfgs.py]: This line takes the minimal alpha across all the directions. This implies that, if the point is near the boundary, the next step is constrained to remain very close and the descent could stall. I expect instead to do different step size in each directions, see 16.71 mentioned in previous comment. Is there another reference for this?

**Comment by vincentqb** [comment]
I will close this PR since 
* #8824 was just approved and implements line search using strong Wolfe method. 
* The changes in reduce.py conflicts with history.
* The box constraint does not appear to be correct, and a different algorithm should be implemented.

File: torch/optim/lbfgs.py (hunk 5 above)
Comment (vincentqb) [reply torch/optim/lbfgs.py]: The relevant part of scipy is the fortran lbfgsb library's [cauchy function](https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb_src/lbfgsb.f#L1209). The gradient is scaled differently for each component, as done in 16.71.
